samza-core/src/main/java/org/apache/samza/zk/ZkJobCoordinator.java

```diff
@@ -554,6 +621,9 @@ public void onBarrierStateChanged(final String version, ZkBarrierForVersionUpgra
       if (ZkBarrierForVersionUpgrade.State.DONE.equals(state)) {
         debounceTimer.scheduleAfterDebounceTime(barrierAction, 0, () -> {
           LOG.info("pid=" + processorId + "new version " + version + " of the job model got confirmed");
+          if (leaderElector.amILeader()) {
```

Comment:
Would be better to add a comment here as to why the leader has to do this state change on end state of the barrier.